### PR TITLE
RPI: Be sure to register Alsa and Pulseaudio.

### PR DESCRIPTION
**Pull request by nsenica**
## Description

Alsa and Pulseaudio were not available in RPi because they were never registered.
This PR attempts to do the same as https://github.com/xbmc/xbmc/pull/13290 but using OptionalsReg.

## How Has This Been Tested?

Compiled and tested in RPI.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by popcornmix** [comment]
Yes, I think this is good.
Most users use the built in AESinkPi, but ALSA is useful for users with USB or I2C soundcards.
Puseaudio is less obviously useful on a Pi - I'm assuming this is for bluetooth audio?

**Comment by nsenica** [comment]
Exactly that :)

**Comment by lrusak** [comment]
Thanks for the PR, however I think you should follow how other window systems have done it. Simply registering all the sinks at once may result in disaster (say pulseaudio is holding the alsa sink).

see, https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/gbm/WinSystemGbm.cpp#L34-L65

**Comment by nsenica** [comment]
That's a good suggestion, will do it and update this PR. 

Thanks for your input.

**Comment by lrusak** [comment]
I'm not sure what @popcornmix thinks but something like this is probably fine

``` c++
  std::string envSink;
  if (getenv("KODI_AE_SINK"))
    envSink = getenv("KODI_AE_SINK");
  if (StringUtils::EqualsNoCase(envSink, "ALSA"))
  {
    OPTIONALS::ALSARegister();
  }
  else if (StringUtils::EqualsNoCase(envSink, "PULSE"))
  {
    OPTIONALS::PulseAudioRegister();
  }
  else
  {
    CAESinkPi::Register();
  }
```

**Comment by nsenica** [comment]
@lrusak That's exactly what I had in mind :) but thanks for the input.

**Comment by diogomsantos** [comment]
@Irusak I don't think that will work. On the Pi (on Raspbian at least, probably OSMC also), besides the own CAESinkPi, for HDMI and Analogue, we also need the Alsa sink, for external DACs and/or bluetooth, and the PulseAudio sink, also for external DACs and/or bluetooth if a user wants to install and use PulseAudio instead. So we need to register all the 3. All the 3 sinks were working on previous Kodi versions.

**Comment by lrusak** [comment]
@diogomsantos you may be able to register the pi and alsa sinks together (like in the old code) but I wouldn't have alsa and pulseaudio together.

see the old code for reference -> https://github.com/xbmc/xbmc/blob/Krypton/xbmc/cores/AudioEngine/AESinkFactory.cpp#L204-L217

**Comment by diogomsantos** [comment]
I know that one can use either Alsa or PulseAudio, not the 2 together at the same time, that is not the issue.
The issue is that now, like the code is, you only have the Pi Sink, so you cannot use ALSA or PulseAudio on the Pi.
My knowledge of C code is very limited, but with your code, you will only have 1 sink again, Alsa, PulseAudio or Pi sink, by this order of detection.

**Comment by lrusak** [comment]
@diogomsantos 
>I know that one can use either Alsa or PulseAudio, not the 2 together at the same time, that is not the issue.
The issue is that now, like the code is, you only have the Pi Sink, so you cannot use ALSA or PulseAudio on the Pi.

So again, the issue is that pulseaudio may be holding onto the ALSA sink, so if kodi tries to use the alsa sink but pulseaudio has exclusive rights to it, then the user will have no sound. 

>My knowledge of C code is very limited, correct me if i am wrong, but with your code, you will only have 1 sink again, Alsa, PulseAudio or Pi sink, by this order of detection.

In the little patch I posted it allows a user (via enviroment variables) to decide which sink type to use. Now we can probably adjust this so that the Pi sink and ALSA sink are used together but not with pulseaudio.

In LE we register all the sinks but this is because for pulseaudio we explicitly disable loading the `module-udev-detect` module so that pulseaudio is only used for bluetooth (or network) audio devices.

So if you want we can probably do this
``` c++
  std::string envSink;
  if (getenv("KODI_AE_SINK"))
    envSink = getenv("KODI_AE_SINK");

  if (StringUtils::EqualsNoCase(envSink, "PULSE"))
  {
    OPTIONALS::PulseAudioRegister();
  }
  else
  {
    OPTIONALS::ALSARegister();
  }

  CAESinkPi::Register();
```

With this patch the Pi sink will always be registered, and unless you explicitly chose PULSE then the ALSA sink will be registered also.

**Comment by diogomsantos** [comment]
I will compile with this latest patch to test. @samnazarko does this looks good for you ?

**Comment by nsenica** [comment]
Thanks for trying it out @diogomsantos 

I'll update the PR accordingly.

**Comment by diogomsantos** [comment]
@nsenica Something like this would be probably better:

`AE::CAESinkFactory::ClearSinks();
 std::string envSink;
 if (getenv("KODI_AE_SINK"))
   envSink = getenv("KODI_AE_SINK");

 if (StringUtils::EqualsNoCase(envSink, "PULSE"))
 {
   OPTIONALS::PulseAudioRegister();
 }
 else
 {
   OPTIONALS::ALSARegister();
 }

CLinuxPowerSyscall::Register();`

This way, the PiSink, which is always needed, is always registered first, and the others, after.

**Comment by nsenica** [comment]
Done :) Thanks..

**Comment by diogomsantos** [comment]
@nsenica Sorry Nuno, doesn't work as intended, don't know why. Please revert to Irusak latest patch, that's good enough. If nobody as any objection, it should be good to merge.

**Comment by popcornmix** [comment]
What order do the audio output devices appear with latest version of this patch?
I think it's important the SinkPi device appears first and is the default as that's what 99% of users should be using.

**Comment by diogomsantos** [comment]
Alsa first, PiSink second. Take a look at my logs above. With my patch, it is the same.

**Comment by popcornmix** [comment]
We need to fix the ordering before merging.
The option that 99% of users want should be the default and be first in list.

**Comment by nsenica** [comment]
As far as I understood, that list is controlled by this map:
https://github.com/xbmc/xbmc/blob/eb161cdb7c4c97e9f054192c9ea9e8d00839c305/xbmc/cores/AudioEngine/AESinkFactory.cpp#L18
which is then retrieved by this function:
https://github.com/xbmc/xbmc/blob/eb161cdb7c4c97e9f054192c9ea9e8d00839c305/xbmc/cores/AudioEngine/AESinkFactory.cpp#L86
The order of which it is iterated here: https://github.com/xbmc/xbmc/blob/eb161cdb7c4c97e9f054192c9ea9e8d00839c305/xbmc/cores/AudioEngine/AESinkFactory.cpp#L90
is based on the key order, since the key is the Sink's name
https://github.com/xbmc/xbmc/blob/eb161cdb7c4c97e9f054192c9ea9e8d00839c305/xbmc/cores/AudioEngine/AESinkFactory.cpp#L22
it means that ALSA will always come first, due to the alphabetical sort.

I might need to investigate the implications, but maybe `m_AESinkRegEntry` can be replaced by a `std::vector` or `std::set` to avoid duplicates - which might have been the reason why this is map - as it is not accessed outside the class nor it is ever accessed by its key. Only to be iterated.

I'll try to figure out in the next few days if I can find a solution for that. I'd appreciate some input from more knowledgeable devs, since this is my first time fiddling with Kodi's core code.

Thanks

**Comment by nsenica** [comment]
Hi @fritsch 

Can you please elaborate on that a bit ?

I have the feeling that what you are suggesting we tried it before (https://github.com/xbmc/xbmc/pull/14453#issuecomment-423184182) and it doesn't work (https://github.com/xbmc/xbmc/pull/14453#issuecomment-423352033) due to the reasons I've explained before, but... I might be wrong :)

Thanks

**Comment by fritsch** [comment]
Replacing it by a simple std::vector with regEntries in there should be good enough.

But this on @FernetMenta to decide. From my look at the code the hashmap is not needed at all. 

Here a POC: http://paste.ubuntu.com/p/rzfymYVhQB/

**Comment by FernetMenta** [comment]
Trying to control the order of displayed items implicitly by an implementation detail is wrong. Never go for the quick hack, do it right. If you want to have influence on the display oder, you need to design for it explicitly.

**Comment by fritsch** [comment]
This is not a quick hack. If one adds entries in order, one expects to get them back in order. If one uses a hashmap and accesses it with an iterator, instead with a key, this is suboptimal design.

**Comment by fritsch** [comment]
Perhaps some more explanation:
- The order is given by the Platform (which is WinSystem for us), cause the Platform (no one else, neither settings, nor AE) know which order is the correct one.
- The only place where sinks are enumerated and classes of sinks are registered are in WinSystem
- Therefore if WinSystem wants to have control on the order of sinks, it needs guarantees of the Public Register API
- This guarantee must be a) documented b) implemented
- For b) it needs a datastructure that implements FIFO

I really don't see a quick hack in that, but a designed API contrackt between AE and the Platforms using it.

**Comment by FernetMenta** [comment]
> For b) it needs a datastructure that implements FIFO

I don't agree because relying on internals is wrong. If priority is something that should be considered or preserved by AE, you need to pass that information explicitly.

Do you really think I made this a map because I could not think of a vector?

**Comment by FernetMenta** [comment]
btw: a similar problem is already solved by renderer. take a look at CLinuxRendererGL::Register(). it registers as "default". this is considered later by CRenderManager::CreateRenderer

**Comment by FernetMenta** [comment]
on the other hand AE does not care about sort order in UI. if this is the only thing you want to solve, you really should do this in ui component.

**Comment by fritsch** [comment]
From the code it is not clear why a map is used. It rather looks like an 1:1 transition of the old Factory we had before. Also looking at the POC: http://paste.ubuntu.com/p/rzfymYVhQB/ <- the advantages of the hashmap are not used at all (direct access to get an element's value).

For the other arguments that you gave: yes, fully agreed. Making it explicit is preferred as it makes the API explicit and not working "by chance" cause of side effects.

**Comment by nsenica** [comment]
@FernetMenta I had a look at CLinuxRendererGL::Register() and CRenderManager::CreateRenderer and I see why the map is helpful, I'm just not sure how we can use this value to influence the order of display.
I totally understand that sort order should be a UI thing and not something we need to change our data structures to accomodate this, but we need something at the UI level that tells us which order to use right ? I believe the default is only a partial solution, as we can bring the default to the top, but it is not A solution.

Can you point me where in the ui components we're picking these sinks up and displaying them on the skin ? I can have a look at it and change the PR accordingly with the proposed changes above.

**Comment by FernetMenta** [comment]
No sure if the requirement has been captured correctly. Why should anybody be interested in the order the UI present the devices to the user? Important is that the desired default device is used on a fresh install. You achieve this by registering the piDevice as "Default" or overriding audiooutput.outputdevice with desired device name.

**Comment by popcornmix** [comment]
While "Default" helps for the first use case, if a user tries getting bluetooth audio working with pulse sink and then wants to switch back to HDMI, they are more likely to choose the first audio output device in the list that appears to work.
And the ALSA sink will appear to work. It just doesn't work as well and isn't recommended.
On Kodi 17 we did have the ALSA sink appearing in audio output device list below the Pi Sink.

**Comment by FernetMenta** [comment]
> On Kodi 17 we did have the ALSA sink appearing in audio output device list below the Pi Sink.

Only by chance and not by design.

Sorting list elements by platform or even system specific rules is a valid and new requirement. Could also be useful for settings like scaling or deinterlacing. I would implement this somewhere in the UI.

**Comment by nsenica** [comment]
So with great help from @diogomsantos, who came up with the idea and tested it, we've added the default audio device for raspberry pi settings. In this case, by the default PI:HDMI will always be used and no other will be picked up just by chance.

I'd like to suggest to merge this PR as is, if you all agree with the changes, and as @FernetMenta suggested, take care of the sorting list as a new requirement, thus a new PR.

Let me know your thoughts.

File: system/settings/rbp2.xml [hunk 1]
```diff
@@ -32,6 +32,11 @@
           <default>30</default> <!-- AE_QUALITY_MID -->
         </setting>
       </group>
+      <group id="1">
```
Comment (lrusak) [inline system/settings/rbp2.xml]: I don't believe you need this change in rbp2.xml as rbp.xml will always be read first. rbp2.xml is just supplemental 

https://github.com/xbmc/xbmc/blob/master/xbmc/settings/Settings.cpp#L550-L553

Comment (nsenica) [reply system/settings/rbp2.xml]: That's a very good observation :) I'll change it right away. Thanks.

**Comment by nsenica** [comment]
Sorry to bother, any change we could get this merged soon ? :)

**Review by lrusak — APPROVED** [review]

**Comment by lrusak** [comment]
needs signoff from @FernetMenta and @popcornmix

**Review by FernetMenta — APPROVED** [review]

**Review by popcornmix — APPROVED** [review]

**Comment by lrusak** [comment]
Thank you @nsenica for you contribution!

**Pull request merged**